pvlib/iotools/bsrn.py

```diff
@@ -4,139 +4,449 @@
+        and '0500'.
+    local_path: str or path-like, optional
+        If specified, path (abs. or relative) of where to save files
+
+    Returns
+    -------
+    data: DataFrame
+        timeseries data from the BSRN archive, see
+        :func:`pvlib.iotools.read_bsrn` for fields. Empty DataFrame if data is
+        not available for the specified period.
+    metadata: dict
+        metadata for the last available monthly file.
+
+    Raises
+    ------
+    KeyError
+        If the specified station does not exist on the FTP server.
 
-    BSRN files are freely available and can be accessed via FTP [3]_. Required
+    Warns
+    -----
+    UserWarning
+        If a requested file is missing a UserWarning is returned with the
+        filename. Also, if no files match the specified station and timeframe.
+
+    Notes
+    -----
+    The username and password for the BSRN FTP server can be obtained for free
+    as described in the BSRN's Data Release Guidelines [3]_.
+
+    Currently only parsing of LR0100, LR0300, and LR0500 is supported. Note
+    not all stations measure LR0300 and LR0500. However, LR0100 is mandatory as
+    it contains the basic irradiance and auxillary measurements. See
+    [4]_ for a description of the different logical records. Future updates may
+    include parsing of additional data and metadata.
+
+    Important
+    ---------
+    While data from the BSRN is generally of high-quality, measurement data
+    should always be quality controlled before usage!
+
+    Examples
+    --------
+    >>> # Retrieve two months irradiance data from the Cabauw BSRN station
+    >>> data, metadata = pvlib.iotools.get_bsrn(  # doctest: +SKIP
+    >>>     start=pd.Timestamp(2020,1,1), end=pd.Timestamp(2020,12,1),   # doctest: +SKIP
+    >>>     station='cab', username='yourusername', password='yourpassword')  # doctest: +SKIP
+
+    See Also
+    --------
+    pvlib.iotools.read_bsrn, pvlib.iotools.parse_bsrn
+
+    References
+    ----------
+    .. [1] `World Radiation Monitoring Center - Baseline Surface Radiation
+        Network (BSRN)
+        <https://bsrn.awi.de/>`_
+    .. [2] `BSRN Data Retrieval via FTP
+       <https://bsrn.awi.de/data/data-retrieval-via-ftp/>`_
+    .. [4] `BSRN Data Release Guidelines
+       <https://bsrn.awi.de/data/conditions-of-data-release/>`_
+    .. [3] `Update of the Technical Plan for BSRN Data Management, 2013,
+       Global Climate Observing System (GCOS) GCOS-174.
+       <https://bsrn.awi.de/fileadmin/user_upload/bsrn.awi.de/Publications/gcos-174.pdf>`_
+    """  # noqa: E501
+    # The FTP server uses lowercase station abbreviations
+    station = station.lower()
 
-    username and password are easily obtainable as described in the BSRN's
-    Data Release Guidelines [4]_.
+    # Generate list files to download based on start/end (SSSMMYY.dat.gz)
+    filenames = pd.date_range(start, end + pd.DateOffset(months=1), freq='1M')\

```

Comment:
When passing `start=end=2019-01-31`, data for both 2019-01 and 2019-02 are returned.  It's a little clunky, but what do you think about `pd.date_range(start.replace(day=1), end + pd.tseries.offsets.MonthEnd(n=0), freq='1M')`?

Reply:
what about just:
`pd.date_range(start.replace(day=1), end.replace(day=1) + pd.DateOffset(months=1), freq='1M')`

I believe you might not even need the `replace()` on start.

Reply:
Works for me.  I wasn't sure about the `replace()` on start, I included it to be safe but I think you're right that it's unnecessary.